### DISCUSSION TITLE
Early JsError: Turbo Module Registry.getEnforcing (...) when trying to build android app using new architecture

### DISCUSSION THREAD

**Discussion by SetRedEyes** [reactwg/react-native-new-architecture#220] [Q&A] [answered]
When app builded successfully i see this Error
Early JsError: Turbo Module Registry.getEnforcing (...): 'PlatformConstants' could not be found. Verify that a module by this name is registered in the native binary. Bridgeless mode: true. Turbo Module interop: true. Modules loaded: ("NativeModules":["TurboModules: "NotFound": ["NativePerformance ObserverCxx","Native PerformanceCxx","PlatformConstants")), 
<img width="1310" alt="SCR-20240904-srom" src="https://github.com/user-attachments/assets/36b49770-6ff5-4ea4-939b-d84be877836f">

>  "devDependencies": {
        "@babel/core": "^7.25.2",
        "@babel/plugin-proposal-nullish-coalescing-operator": "^7.18.6",
        "@babel/plugin-proposal-optional-chaining": "^7.21.0",
        "@babel/preset-env": "^7.25.4",
        "@babel/runtime": "^7.25.6",
        "@react-native/eslint-config": "0.75.2",
        "@react-native/metro-config": "0.75.2",
        "@react-native/typescript-config": "0.75.2",
        "@types/react": "^18.3.5",
        "@types/react-native-background-timer": "^2.0.2",
        "@types/react-native-dotenv": "^0.2.2",
        "@types/react-test-renderer": "^18.3.0",
        "babel-jest": "^29.7.0",
        "babel-plugin-module-resolver": "^5.0.2",
        "babel-plugin-transform-remove-console": "^6.9.4",
        "eslint": "^9.9.1",
        "jest": "^29.7.0",
        "prettier": "^3.3.3",
        "react-native-dotenv": "^3.4.11",
        "react-native-flipper-performance-plugin": "^0.4.0",
        "react-test-renderer": "18.3.1",
        "typescript": "5.5.4"
    },
    "dependencies": {
        "@expo/config": "^9.0.3",
        "@expo/env": "^0.3.0",
        "@hookform/resolvers": "^3.9.0",
        "@react-native-async-storage/async-storage": "^2.0.0",
        "@react-native-community/checkbox": "^0.5.17",
        "@react-native-community/hooks": "^3.0.0",
        "@react-native/codegen": "0.75.2",
        "@react-native/metro-config": "^0.75.2",
        "@react-navigation/native": "^6.1.18",
        "@react-navigation/native-stack": "^6.11.0",
        "@react-three/fiber": "^8.17.6",
        "@reduxjs/toolkit": "^2.2.7",
        "@sentry/react-native": "^5.31.1",
        "@shopify/react-native-skia": "^1.3.11",
        "@types/lodash": "^4.17.7",
        "@types/three": "^0.168.0",
        "axios": "^1.7.7",
        "base64-arraybuffer": "^1.0.2",
        "dayjs": "^1.11.13",
        "expo": "51.0.32",
        "expo-asset": "^10.0.10",
        "expo-av": "~14.0.7",
        "expo-constants": "~16.0.2",
        "expo-file-system": "~17.0.1",
        "expo-font": "~12.0.10",
        "expo-gl": "~14.0.2",
        "expo-modules-core": "^1.12.24",
        "expo-screen-orientation": "~7.0.5",
        "expo-sqlite": "~14.0.6",
        "immer": "^10.1.1",
        "lodash": "^4.17.21",
        "react": "18.3.1",
        "react-hook-form": "^7.53.0",
        "react-icomoon": "^2.5.7",
        "react-native": "0.75.2",
        "react-native-background-timer": "^2.4.1",
        "react-native-ble-manager": "^11.5.5",
        "react-native-confirmation-code-field": "^7.4.0",
        "react-native-fs": "^2.20.0",
        "react-native-geolocation-service": "^5.3.1",
        "react-native-gesture-handler": "~2.19.0",
        "react-native-just-checkbox": "^2.0.15",
        "react-native-logs": "^5.1.0",
        "react-native-material-menu": "^2.0.0",
        "react-native-radio-buttons-group": "^3.1.0",
        "react-native-reanimated": "3.15.1",
        "react-native-redash": "^18.1.3",
        "react-native-safe-area-context": "^4.10.9",
        "react-native-screens": "~3.34.0",
        "react-native-svg": "^15.6.0",
        "react-native-svg-transformer": "^1.5.0",
        "react-native-system-setting": "^1.7.6",
        "react-native-toast-notifications": "^3.4.0",
        "react-native-ui-datepicker": "^2.0.3",
        "react-native-walkthrough-tooltip": "^1.6.0",
        "react-native-wifi-reborn": "^4.12.1",
        "react-redux": "^9.1.2",
        "rn-animated-ellipsis": "^2.1.3",
        "three": "0.168.0",
        "three-stdlib": "^2.32.2",
        "victory-native": "^41.1.1",
        "yup": "^1.4.0",
        "zustand": "^4.5.5"
    },

**Comment by cortinico** [comment]
> Early JsError: Turbo Module Registry.getEnforcing (...): 'PlatformConstants' could not be found

This happens when your app is trying to access `NativeModules.PlatformConstants` but that module is not found.

In this case it generally means there is something going on with your application during the startup. Without a reproducer we can't help you further.

**Reply by SetRedEyes** [reply]
Hi, @cortinico!
Thank you for your fast answer!

Here is link on the project with error which was made with reproducer template:
https://github.com/SetRedEyes/new_arch_native_modules_error

Thkank you for your attention!

**Comment by cortinico** [comment, marked as answer]
Hey @SetRedEyes Thanks for providing a reproducer.

So the issue you're seeing is because your application is missing the `libappmodule.so`. That is the default dynamic library used by your app to instantiate React Native.

When you create a new application, we create a default `libappmodule.so` for you with all the necessary bits needed to load React Native.

However here you're overriding it by providing your own `CMakeLists.txt` file as I see you're compiling some custom C++ code. You can provide your own CMake file, but you need to make sure the libappmodules.so is still produced.

So you will have to do the following.

1. Update your CMake file here as follows:
```cmake

# For more information about using CMake with Android Studio, read the
# documentation: https://d.android.com/studio/projects/add-native-code.html.
# For more examples on how to use CMake, see https://github.com/android/ndk-samples.

# Sets the minimum CMake version required for this project.
cmake_minimum_required(VERSION 3.22.1)

# Declares the project name. The project name can be accessed via ${ PROJECT_NAME},
# Since this is the top level CMakeLists.txt, the project name is also accessible
# with ${CMAKE_PROJECT_NAME} (both CMake variables are in-sync within the top level
# build script scope).
project(appmodules)

# This file includes all the necessary to let you build your application with the New Architecture.
include(${REACT_ANDROID_DIR}/cmake-utils/ReactNative-application.cmake)

# Creates and names a library, sets it as either STATIC
# or SHARED, and provides the relative paths to its source code.
# You can define multiple libraries, and CMake builds them for you.
# Gradle automatically packages shared libraries with your APK.
#
# In this top level CMakeLists.txt, ${CMAKE_PROJECT_NAME} is used to define
# the target library name; in the sub-module's CMakeLists.txt, ${PROJECT_NAME}
# is preferred for the same purpose.
#
# In order to load a library into your app from Java/Kotlin, you must call
# System.loadLibrary() and pass the name of the library defined here;
# for GameActivity/NativeActivity derived applications, the same library name must be
# used in the AndroidManifest.xml file.
add_library(Cortex_C_Code_V1 SHARED
        Cortex_C_Code_V1.cpp
        Cortex_PPG_EEG.cpp
        Cortex_PPG_EEG.h
        Cortex_PPG_EEG_data.cpp
        Cortex_PPG_EEG_private.h
        Cortex_PPG_EEG_types.h
        pch.cpp
        pch.h
        rt_nonfinite.cpp
        rt_nonfinite.h
        rtGetInf.cpp
        rtGetInf.h
        rtGetNaN.cpp
        rtGetNaN.h
        rtwtypes.h)

# Specifies libraries CMake should link to your target library. You
# can link libraries from various origins, such as libraries defined in this
# build script, prebuilt third-party libraries, or Android system libraries.
target_link_libraries(Cortex_C_Code_V1
    # List libraries link to the target library
    android
    log)
target_link_libraries(appmodules Cortex_C_Code_V1)
```

2. Add the default OnLoad.cpp file for libappmodules.so defined here: https://github.com/facebook/react-native/blob/main/packages/react-native/ReactAndroid/cmake-utils/default-app-setup/OnLoad.cpp

That file will have to live in the same folder of your CMakeLists.txt file.

Once you have those two changes, you'll be able to build and run correctly.

**Reply by SetRedEyes** [reply]
Thank you!!!
